File: xion-android-demo/app/src/main/java/com/burnt/xiondemo/data/remote/XionOAuthApi.kt
package com.burnt.xiondemo.data.remote

import com.burnt.xiondemo.data.model.OAuthTokens
import com.burnt.xiondemo.data.model.OAuthUserInfo
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.POST

interface XionOAuthApi {

    @FormUrlEncoded
    @POST("oauth2/token")
    suspend fun exchangeCode(
        @Field("grant_type") grantType: String = "authorization_code",
        @Field("code") code: String,
        @Field("redirect_uri") redirectUri: String,
        @Field("code_verifier") codeVerifier: String,
        @Field("client_id") clientId: String
    ): OAuthTokens

    @FormUrlEncoded
    @POST("oauth2/token")
    suspend fun refreshToken(
        @Field("grant_type") grantType: String = "refresh_token",
        @Field("refresh_token") refreshToken: String,
        @Field("client_id") clientId: String
    ): OAuthTokens

    @GET("api/v1/me")
    suspend fun getUserInfo(
        @Header("Authorization") authorization: String
    ): OAuthUserInfo
}

File: xion-ios/XionDemo/Data/Remote/XionOAuthAPI.swift
import Foundation

final class XionOAuthAPI {

    private let baseUrl: String

    init(baseUrl: String = Constants.restUrl) {
        self.baseUrl = baseUrl
    }

    func exchangeCode(
        code: String,
        codeVerifier: String,
        clientId: String
    ) async throws -> OAuthTokens {
        let url = URL(string: "\(baseUrl)oauth2/token")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        let body = [
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": Constants.oauthRedirectUri,
            "code_verifier": codeVerifier,
            "client_id": clientId
        ]

        request.httpBody = body
            .map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? $0.value)" }
            .joined(separator: "&")
            .data(using: .utf8)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw APIError.requestFailed(statusCode: (response as? HTTPURLResponse)?.statusCode ?? 0)
        }

        return try JSONDecoder().decode(OAuthTokens.self, from: data)
    }

    func refreshToken(
        refreshToken: String,
        clientId: String
    ) async throws -> OAuthTokens {
        let url = URL(string: "\(baseUrl)oauth2/token")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        let body = [
            "grant_type": "refresh_token",
            "refresh_token": refreshToken,
            "client_id": clientId
        ]

        request.httpBody = body
            .map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? $0.value)" }
            .joined(separator: "&")
            .data(using: .utf8)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw APIError.requestFailed(statusCode: (response as? HTTPURLResponse)?.statusCode ?? 0)
        }

        return try JSONDecoder().decode(OAuthTokens.self, from: data)
    }

    func getUserInfo(accessToken: String) async throws -> OAuthUserInfo {
        let url = URL(string: "\(baseUrl)api/v1/me")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw APIError.requestFailed(statusCode: (response as? HTTPURLResponse)?.statusCode ?? 0)
        }

        return try JSONDecoder().decode(OAuthUserInfo.self, from: data)
    }
}

enum APIError: LocalizedError {
    case requestFailed(statusCode: Int)

    var errorDescription: String? {
        switch self {
        case .requestFailed(let code): return "Request failed with status \(code)"
        }
    }
}
